Src/Plugins/Security/PFXCryptographicKeyFileValidator.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    internal static class PFXCryptographicKeyFileValidator
+    {
+#pragma warning disable IDE0060 // Remove unused parameter
+        public static string IsValid(
+            ref string matchedPattern,
+            ref Dictionary<string, string> groups,
+            ref bool performDynamicValidation,
+            ref string failureLevel,
+            ref string fingerprint)
+        {
+#pragma warning restore IDE0060
+            fingerprint = $"[cert={matchedPattern}]";
+
+            // This plugin does not perform any dynamic validation.
+            // We therefore set this setting to false. This is a
+            // clue to the caller not to warn the user that, e.g.,
+            // dynamic analysis was available but not exercised.
+            performDynamicValidation = false;
+
+            return TryLoadCertificate(matchedPattern).ToString();
+        }
+
+        private static ValidationState TryLoadCertificate(string certificatePath)
+        {
+            X509Certificate2 certificate = null;
+            try
+            {
+                // If this certificate needs a password or it is a bundle, it will throw an exception.
+                certificate = new X509Certificate2(certificatePath);
+
+                return certificate.PrivateKey != null ? ValidationState.Authorized : ValidationState.NoMatch;
+            }
+            catch (CryptographicException ex)
+            {
+                return ex.Message switch
+                {
+                    "Cannot find the original signer." => TryLoadCertificateCollection(certificatePath),
+                    "The specified network password is not correct." => ValidationState.Unknown,
+                    _ => ValidationState.Unknown,
+                };
+            }
+            catch (Exception)
+            {
+                return ValidationState.Unknown;
```

Comment:
>ValidationState.Unknown [](start = 23, length = 23)

see ValidatorBase for our standard helper to generate a return value based on an exception.

CreateReturnValueForException. It looks like that helper might need reworking or we need to create a second helper for this scenario. We shouldn't be returning HostUnknown here, probably just Unknown. We *do* want to emit the public key token to make clear the identity of the thing under analysis.

Reply:
for the public token, if the file is a bundle, what would be our public token? the concat of all?

---
In reply to: [554470932](https://github.com/microsoft/sarif-pattern-matcher/pull/96#discussion_r554470932) [](ancestors = 554470932)

Reply:
Another question: today we have the enum "Authorized", but this validator should return something like "Match" or something related. Don't know

---
In reply to: [554471205](https://github.com/microsoft/sarif-pattern-matcher/pull/96#discussion_r554471205) [](ancestors = 554471205,554470932)

Reply:
added a few static methods in validatorbase which has a file in the name, so we can distinguish between content vs file

---
In reply to: [554471302](https://github.com/microsoft/sarif-pattern-matcher/pull/96#discussion_r554471302) [](ancestors = 554471302,554471205,554470932)